File: iOS/ssabree/ssabree/Features/Board/BoardEditViewModel.swift
import Foundation
import Observation

// MARK: - UI State

struct BoardEditUiState {
    var postId: Int = 0
    var title: String = ""
    var content: String = ""
    var isLoading: Bool = false
    var isSubmitting: Bool = false
    var isSubmitSuccess: Bool = false
    var submitError: String? = nil

    var isSubmitEnabled: Bool {
        !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
        !content.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
        !isSubmitting
    }
}

// MARK: - ViewModel

@Observable
@MainActor
final class BoardEditViewModel {
    private let postRepository: PostRepository

    var uiState = BoardEditUiState()

    init(postRepository: PostRepository) {
        self.postRepository = postRepository
    }

    // MARK: - Load Post

    func loadPost(postId: Int) async {
        uiState.postId = postId
        uiState.isLoading = true

        let result = await postRepository.getPostDetail(postId: postId)

        switch result {
        case .success(let detail):
            uiState.title = detail.title
            uiState.content = detail.content
        case .failure(let error):
            print("Failed to load post: \(error)")
            uiState.submitError = "게시글을 불러오는데 실패했습니다."
        }

        uiState.isLoading = false
    }

    // MARK: - Title & Content

    func onTitleChange(_ value: String) {
        uiState.title = value
    }

    func onContentChange(_ value: String) {
        uiState.content = value
    }

    // MARK: - Submit

    func onSubmit() async {
        guard uiState.isSubmitEnabled else { return }

        uiState.isSubmitting = true
        uiState.submitError = nil

        let updateInfo = PostUpdateInfo(
            title: uiState.title.trimmingCharacters(in: .whitespacesAndNewlines),
            content: uiState.content.trimmingCharacters(in: .whitespacesAndNewlines)
        )

        let result = await postRepository.updatePost(postId: uiState.postId, post: updateInfo)

        switch result {
        case .success:
            uiState.isSubmitSuccess = true
        case .failure(let error):
            print("Failed to update post: \(error)")
            uiState.submitError = "게시글 수정에 실패했습니다."
        }

        uiState.isSubmitting = false
    }

    func clearSubmitError() {
        uiState.submitError = nil
    }

    func clearSubmitSuccess() {
        uiState.isSubmitSuccess = false
    }
}

File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/board/screen/BoardEditViewModel.kt
package com.ssafy.ssabree.features.board.screen

import android.content.Context
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.repository.PostRepository
import com.ssafy.ssabree.core.repository.model.PostUpdateInfo
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

private const val TAG = "BoardEditViewModel"

data class BoardEditUiState(
    val postId: Long = 0L,
    val title: String = "",
    val content: String = "",
    val isLoading: Boolean = false,
    val isSubmitting: Boolean = false,
    val isSubmitSuccess: Boolean = false,
    val submitError: String? = null,
) {
    val isSubmitEnabled: Boolean
        get() = title.isNotBlank() &&
            content.isNotBlank() &&
            !isSubmitting
}

class BoardEditViewModel(
    private val postRepository: PostRepository,
    private val appContext: Context
) : ViewModel() {

    private val _uiState = MutableStateFlow(BoardEditUiState())
    val uiState: StateFlow<BoardEditUiState> = _uiState.asStateFlow()

    fun loadPost(postId: Long) {
        _uiState.update { it.copy(postId = postId, isLoading = true) }
        viewModelScope.launch {
            postRepository.getPostDetail(postId)
                .onSuccess { detail ->
                    Log.d(TAG, "loadPost: succeed")
                    _uiState.update {
                        it.copy(
                            postId = postId,
                            title = detail.title,
                            content = detail.content,
                            isLoading = false
                        )
                    }
                }
                .onFailure { e ->
                    Log.d(TAG, "loadPost: failed (${e.message})")
                    _uiState.update { it.copy(isLoading = false, submitError = e.message) }
                }
        }
    }

    fun onTitleChange(value: String) {
        _uiState.update { it.copy(title = value) }
    }

    fun onContentChange(value: String) {
        _uiState.update { it.copy(content = value) }
    }

    fun onSubmit() {
        if (!_uiState.value.isSubmitEnabled) return

        _uiState.update { it.copy(isSubmitting = true, submitError = null) }
        viewModelScope.launch {
            val state = _uiState.value

            postRepository.updatePost(
                state.postId,
                PostUpdateInfo(
                    title = state.title,
                    content = state.content
                )
            ).onFailure { e ->
                Log.d(TAG, "onSubmit: failed (${e.message})")
                _uiState.update {
                    it.copy(
                        isSubmitting = false,
                        submitError = e.message ?: "알 수 없는 오류가 발생했습니다."
                    )
                }
            }.onSuccess {
                Log.d(TAG, "onSubmit: succeed")
                _uiState.update {
                    it.copy(
                        isSubmitting = false,
                        isSubmitSuccess = true
                    )
                }
            }
        }
    }

    fun clearSubmitError() {
        _uiState.update { it.copy(submitError = null) }
    }

    fun clearSubmitSuccess() {
        _uiState.update { it.copy(isSubmitSuccess = false) }
    }
}
